internal/client/consensus/grandpa/environment.go

```diff
@@ -625,3 +630,1018 @@ type votePrecommit[H runtime.Hash, N runtime.Number] struct {
+// justification from a commit (when validating). Justifications for blocks that change the authority set will always
+// be processed, otherwise we'll only process justifications if the last one was justificationPeriod blocks ago.
+func shouldProcessJustification[
+	H runtime.Hash,
+	N runtime.Number,
+	Hasher runtime.Hasher[H],
+	Header runtime.Header[N, H],
+	E runtime.Extrinsic,
+](
+	client ClientForGrandpa[H, N, Hasher, Header, E],
+	justificationPeriod uint32,
+	number N,
+	enactsChange bool,
+) bool {
+	if enactsChange {
+		return true
+	}
+
+	lastFinalizedNumber := client.Info().FinalizedNumber
+
+	// keep the first justification before reaching the justification period
+	if lastFinalizedNumber == 0 {
+		return true
+	}
+
+	return lastFinalizedNumber/N(justificationPeriod) != number/N(justificationPeriod)
+}
+
+type justificationOrCommit interface {
+	isJustificationOrCommit()
+}
+
+type justificationOrCommitJustification[H runtime.Hash, N runtime.Number, Header runtime.Header[N, H]] struct {
+	GrandpaJustification[H, N, Header]
+}
+
+func (justificationOrCommitJustification[H, N, Header]) isJustificationOrCommit() {}
+
+type justificationOrCommitCommit[H runtime.Hash, N runtime.Number] struct {
+	primitives.RoundNumber
+	grandpa.Commit[H, N, primitives.AuthoritySignature, primitives.AuthorityID]
+}
+
+func (justificationOrCommitCommit[H, N]) isJustificationOrCommit() {}
+
+// Finalize the given block and apply any authority set changes. If an authority set change is enacted then a
+// justification is created (if not given) and stored with the block when finalizing it. This method assumes that
+// the block being finalized has already been imported.
+func finalizeBlock[
+	H runtime.Hash,
+	N runtime.Number,
+	Hasher runtime.Hasher[H],
+	Header runtime.Header[N, H],
+	E runtime.Extrinsic,
+](
+	client ClientForGrandpa[H, N, Hasher, Header, E],
+	authoritySet *SharedAuthoritySet[H, N],
+	justificationGenerationPeriod *uint32,
+	hash H,
+	number N,
+	justificationOrCommit justificationOrCommit,
+	initialSync bool,
+	justificationSender *GrandpaJustificationSender[H, N, Header], // can be nil
+) error {
+
+	// NOTE: lock must be held through writing to DB to avoid race. this lock also implicitly synchronises the check
+	// for last finalized number below.
+	authoritySet.mtx.Lock()
+	defer authoritySet.mtx.Unlock()
+
+	status := client.Info()
+
+	if number <= status.FinalizedNumber {
+		hash, err := client.Hash(number)
+		if err != nil {
+			return err
+		}
+		if hash != nil {
+			// This can happen after a forced change (triggered manually from the runtime when finality is stalled),
+			// since the voter will be restarted at the median last finalized block, which can be lower than the local
+			// best finalized block.
+			logger.Warnf("Re-finalized block %s (%d) in the canonical chain, current best finalized is %d",
+				hash,
+				number,
+				status.FinalizedNumber,
+			)
+			return nil
+		}
+	}
+
+	// TODO: do I need to clone this?
```

Comment:
Still valid todo?

Reply:
Yes I'd like to keep this.  After doing some more investigation I do think I need to deep copy the `AuthoritySet` attributes.